wgpu-hal/src/vulkan/device.rs

```diff
@@ -1863,12 +1863,23 @@ impl crate::Device for super::Device {
                 .build()
         }];
 
+        let vertex_cache = desc.vertex_stage.cache.map(|it| it.raw);
+        let fragment_cache = desc
+            .fragment_stage
+            .as_ref()
+            .and_then(|it| it.cache)
+            .map(|it| it.raw);
+        // TODO: What should the behaviour be when both are set and different?
```

Comment:
We should just assert they're the same, and document the requirement on the wgpu-hal interface - wgpu-core will always uphold it, so this is fine.

Reply:
How should `wgpu-core` uphold it? The way that it works with `PipelineCompilationOptions` works, it's specified directly on each of the vertex and fragment shaders.

The problem becomes that `constants` are per-shader, whereas the pipeline cache is across the entire pipeline (at least *in Vulkan*).
I think the least bad option is to specify that the vertex cache will be prioritised on platforms where these are the same?

Reply:
Ah! Core should check to make sure the caches are the same - honestly it sounds like it needs to be part of the base pipeline descriptor instead of a per-entrypoint argument.

Reply:
Yeah, possibly adding it to the root descriptor is best.

But it's a bit annoying to ask every single user to need to add this field, when they already have a defaulted `PipelineCompilationOptions`.

I can also imagine these being meaningfully distinct, such as for producing DXIL (although combining them would also work in that case).

An alternative is to merge the caches on Vulkan, and make them alias the same raw cache object. There don't seem to be any good options, because of the semantics here

I won't be addressing this until at least Thursday, but I suppose putting in the top-level descriptor is actually best.

Hinting to users in the upgrade that the cache exists and that they should/could use it is possibly a good thing, anyway

Reply:
Sounds like root is the way to go. Give me a buzz when that's ready and we can get this in.